r/tests/testthat/test-data-type.R

```diff
@@ -386,20 +386,34 @@ test_that("DictionaryType validation", {
 
 test_that("decimal type and validation", {
   expect_r6_class(decimal(4, 2), "Decimal128Type")
+  expect_r6_class(decimal(39, 2), "Decimal256Type")
 
-  expect_error(decimal("four"), '"precision" must be an integer')
-  expect_error(decimal(4, "two"), '"scale" must be an integer')
-  expect_error(decimal(NA, 2), '"precision" must be an integer')
-  expect_error(decimal(4, NA), '"scale" must be an integer')
+  expect_error(decimal("four"), "`precision` must be an integer")
+  expect_error(decimal(4, "two"), "`scale` must be an integer")
+  expect_error(decimal(NA, 2), "`precision` must be an integer")
+  expect_error(decimal(4, NA), "`scale` must be an integer")
+
+  # decimal() creates either decimal128 or decimal256 based on precision
+  expect_identical(class(decimal(38, 2)), class(decimal128(38, 2)))
+  expect_identical(class(decimal(39, 2)), class(decimal256(38, 2)))
 
-  # decimal() is just an alias for decimal128() for backwards compatibility
   expect_r6_class(decimal128(4, 2), "Decimal128Type")
-  expect_identical(class(decimal(2, 4)), class(decimal128(2, 4)))
 
-  expect_error(decimal128("four"), '"precision" must be an integer')
-  expect_error(decimal128(4, "two"), '"scale" must be an integer')
-  expect_error(decimal128(NA, 2), '"precision" must be an integer')
-  expect_error(decimal128(4, NA), '"scale" must be an integer')
+  expect_error(decimal128("four"), "`precision` must be an integer")
+  expect_error(decimal128(4, "two"), "`scale` must be an integer")
+  expect_error(decimal128(NA, 2), "`precision` must be an integer")
+  expect_error(decimal128(4, NA), "`scale` must be an integer")
+  expect_error(decimal128(3:4, NA), "`precision` must have size 1. not size 2")
+  expect_error(decimal128(4, 2:3), "`scale` must have size 1. not size 2")
+
+  expect_r6_class(decimal256(4, 2), "Decimal256Type")
+
+  expect_error(decimal256("four"), "`precision` must be an integer")
+  expect_error(decimal256(4, "two"), "`scale` must be an integer")
+  expect_error(decimal256(NA, 2), "`precision` must be an integer")
+  expect_error(decimal256(4, NA), "`scale` must be an integer")
+  expect_error(decimal256(3:4, NA), "`precision` must have size 1. not size 2")
+  expect_error(decimal256(4, 2:3), "`scale` must have size 1. not size 2")
```

Comment:
@pitrou we discussed about removing those test since the error in handled and tested in C++
see @jonkeane's comment: https://github.com/apache/arrow/pull/11790#discussion_r758496846
> If the c++ errors are tested in the c++ code, we don't need to test them here as well.

Reply:
I've been looking for the c++ tests for these errors, and I haven't seen any. There _are_ glib tests: https://github.com/cyb70289/arrow/blob/994074d2e7ff073301e0959dbc5bb595a1e2a41b/c_glib/test/test-decimal256-data-type.rb#L49-L54

And the PR that made these messages slightly nicer doesn't change any c++ tests: https://github.com/apache/arrow/pull/11773/files

Am I missing something, or are we not testing those in c++?

Reply:
apologies. I was under the impression the c_glib _are_ C++ tests

https://github.com/apache/arrow/blob/2baed02d259fe3534d20de14af6f7df48611b858/c_glib/test/test-decimal256-data-type.rb#L51

Reply:
I will add the tests back in, both for `decimal128()` and `decimal256()`.

Reply:
That's good for now, though I wonder if we should make a Jira for having them tested in C++ (which AFAIK is a better place for tests of that kind of functionality)

Reply:
Done! Thanks both.

Reply:
I sent the message above before seeing your comment @jonkeane. @pitrou what do you think? 
2 Jiras:
1. [C++] add tests
2. [R] once the C++ tests are in place remove the R tests

Reply:
ARROW-15162 - to add C++ unit tests

Reply:
Let's do (1), and then add a TODO note to the tests here that reference that ticket

Reply:
Ooops, the GH interface hadn't updated before I sent that, thanks for the Jira!

Reply:
Done (added TODO notes)